### DISCUSSION TITLE
Need clarification in single dimension matrix multiplication

### DISCUSSION THREAD

**Discussion by Anshul007** [mrdbourke/pytorch-deep-learning#1107] [Q&A]
Hi There, 

I have always seen any coding language from the math's perspective. but I got one doubt in the video [Learn PyTorch for deep learning in a day. Literally](https://www.youtube.com/watch?v=Z_ikDlimN6A&t=2732s) at `2:22:05`

1. shape of the tensor
```
sample = torch.tensor([1,2,3],[4,5,6],[7,8,9])
sample.shape
```
output :
`torch.Size([3, 3])`
> This is for me is this is matrix of size/shape is 3 by 3. i.e. 3 rows and 3 columns
2.  Now matrix multiplication of 2 different matrix
```
one = torch.tensor([[8, 17, 21],
                    [9, 20, 12],
                    [16, 15, 23]])
two = torch.tensor([[14, 7, 22],
                    [10, 18, 19],
                     [25, 6, 11]])
torch.matmul(one,two)
```
output :
```
tensor([[807, 488, 730],
        [626, 495, 710],
        [949, 520, 890]])
```
> The math behind is for [807 = 8\*14 + 17\*10 + 21\*25]. same also explained in lecture [How to Multiply Matrices](https://www.mathsisfun.com/algebra/matrix-multiplying.html)
i.e. in math as well as in code, for any matrix multiplication if matrix `one` is 2 by 3 and matrix `two` is 3 by 4 then the resulted matrix will be 2 by 4. because this will take first row from left side matrix and first column from second matrix then do a multiply and store it in first place in the resulted matrix.

------
Final question
3. Then why multiplication for single dimension is different for code wise and math wise.
```
one = torch.tensor([5,1,2])
two = torch.tensor([3,0,4])
torch.matmul(one,two)
``` 
output:
```
tensor(23)
```
In Math
`one(1by3) matmul two(1by3) is not possible. Cause, in-order to do multiplication rows from the left matrix has to be same as columns from the right matrix` 

**Note:** Let me know if I am not clear.

**Comment by sachinlodhi** [comment]
Okay let us start with the **first** part:
`sample = torch.tensor([1,2,3],[4,5,6],[7,8,9])
sample.shape`
It will throw syntax error as you are passing three lists as an arguments separated by comma.
The correct way to do it is:
`sample = torch.tensor([[1,2,3],[4,5,6],[7,8,9]])
sample.shape`
now it will be 
`torch.Size([3, 3])`
Regarding your **second** question:
if both the tensors are 1D(as in your case) then it will return scaler value(23) in your case.
It is mentioned in the documentation here: https://pytorch.org/docs/stable/generated/torch.matmul.html#:~:text=The%20behavior%20depends,product%20is%20returned.
